### PR TITLE
Add local registry feature

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
Signed-off-by: Prajyot Parab <prajyot.parab@ibm.com>

### CODE REVIEW

File: tasks/setup_registry.yaml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    shell: |
+      registry_token=`echo -n "admin:admin" | base64 -w0`
+      jq '.auths += {"{{ local_registry }}": {"auth": "'$registry_token'","email": "noemail@localhost"}}' < ~/pull-secret.json > ~/pull-secret-updated.json
+
+  - name: Mirror the registry
+    shell: oc adm -a ~/pull-secret-updated.json release mirror \
+      --from={{ release_image }} \
+      --to={{ local_registry }}/{{ local_repo }} \
+      --to-release-image={{ local_registry }}/{{ local_repo }}:{{ ocp_release }}
+    register: registry
```
Comment (christianh814) [inline tasks/setup_registry.yaml:97]: I would use a var here. Not everyone will have access to the internet if they are in a disconnected situation. I still like the idea and would keep it though!

Maybe add in the vars...

```
setup_registry:
  deploy: true
  autosync_registry: true
```

Then do something like ...

```
when: setup_registry.autosync_registry
```

The default is to set `setup_registry.autosync_registry` as false

File: vars/main.yml [hunk 2]
```diff
@@ -11,4 +11,13 @@ ocp_installer: "https://mirror.openshift.com/pub/openshift-v4/clients/ocp/4.4.3/
...
+  enabled: false
+setup_registry: false
+registry_image: docker.io/library/registry:2
+registry_host: "registry.{{ dns.clusterid }}.{{ dns.domain | lower  }}"
+local_registry: "{{ registry_host }}:5000"
+local_repo: "ocp4/openshift4"
+product_repo: "openshift-release-dev"
+release_name: "ocp-release"
+ocp_release: "4.4.9-x86_64"
+release_image: "quay.io/{{ product_repo }}/{{ release_name }}:{{ ocp_release }}"
```
Comment (christianh814) [inline vars/main.yml:23]: Again, see my comment about structure.

Comment (Prajyot-Parab) [reply vars/main.yml]: Done

File: tasks/setup_registry.yaml (hunk 1 above)
Comment (Prajyot-Parab) [reply tasks/setup_registry.yaml]: Done

File: tasks/setup_registry.yaml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  retries: 3
+  delay: 30
+
+- name: Mirror the registry
+  when: pod_state.stdout != ""
+  block:
+  - name: Generate the base64-encoded user name and password or token for your mirror registry
+    shell: |
+      registry_token=`echo -n "admin:admin" | base64 -w0`
+      jq '.auths += {"{{ local_registry }}": {"auth": "'$registry_token'","email": "noemail@localhost"}}' < ~/pull-secret.json > ~/pull-secret-updated.json
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply tasks/setup_registry.yaml:104]: Done

File: tasks/setup_registry.yaml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  with_items:
+    - auth
+    - certs
+    - data
+
+- name: Create self signed certificate for local registry
+  shell: openssl req -newkey rsa:4096 -nodes -sha256 -keyout domain.key -x509 -days 365 -out domain.crt \
+          -subj "/C=US/ST=State/L=City/O=Company, Inc./OU=Department Name/CN={{ registry_host }}"
+  args:
+    chdir: /opt/registry/certs
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply tasks/setup_registry.yaml]: Done

File: docs/examples/vars-local-registry-ppc64le.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    macaddr: "52:54:00:8e:10:34"
+other:
+  - name: "non-cluster-vm"
+    ipaddr: "192.168.7.31"
+    macaddr: "52:54:00:f4:2e:2e"
+
+ppc64le: true
+setup_registry: true
+registry_image: docker.io/ibmcom/registry-ppc64le:2.6.2.5
+ocp_release: "4.4.9-ppc64le"
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply docs/examples/vars-local-registry-ppc64le.yaml]: Done

File: docs/examples/vars-local-registry.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  - name: "non-cluster-vm"
+    ipaddr: "192.168.7.31"
+    macaddr: "52:54:00:f4:2e:2e"
+
+setup_registry: true
+ocp_bios: "https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.4/4.4.3/rhcos-4.4.3-x86_64-metal.x86_64.raw.gz"
+ocp_initramfs: "https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.4/4.4.3/rhcos-4.4.3-x86_64-installer-initramfs.x86_64.img"
+ocp_install_kernel: "https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.4/4.4.3/rhcos-4.4.3-x86_64-installer-kernel-x86_64"
+ocp_client: "https://mirror.openshift.com/pub/openshift-v4/clients/ocp/4.4.3/openshift-client-linux-4.4.3.tar.gz"
+ocp_installer: "https://mirror.openshift.com/pub/openshift-v4/clients/ocp/4.4.3/openshift-install-linux-4.4.3.tar.gz"
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply docs/examples/vars-local-registry.yaml]: Done

File: docs/vars-doc.md [hunk 7]
```diff
@@ -231,6 +231,36 @@ other:
...
+product_repo: "openshift-release-dev"
+release_name: "ocp-release"
+ocp_release: "4.4.9-x86_64"
+```
+
+* `registry_image` - This is the name of the image used for creating registry container.
+* `local_repo` - This is the name of the repo in your registry.
+* `product_repo` - Where the images are hosted in the product repo.
+* `release_name` - This is the name of the image release.
+* `ocp_release` - The the version of OpenShift you want to sync.
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply docs/vars-doc.md]: Done

File: tasks/setup_registry.yaml [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        service: https
+        permanent: yes
+        state: enabled
+
+- name: Restarting regitry services
+  service:
+    name: "{{ item }}"
+    state: restarted
+  with_items:
+    - "{{ registry_services }}"
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply tasks/setup_registry.yaml:89]: Those services need to be restarted at that point itself before continuing with rest tasks, so keeping as it is.

File: tasks/setup_registry.yaml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        state: enabled
+      with_items:
+      - internal
+      - public
+
+    - name: Add http service
+      firewalld:
+        service: https
+        permanent: yes
+        state: enabled
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply tasks/setup_registry.yaml]: Done

File: tasks/setup_registry.yaml [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+- name: Copy self-signed certificate
+  copy:
+    src: /opt/registry/certs/domain.crt
+    dest: /etc/pki/ca-trust/source/anchors/domain.crt
+    remote_src: yes
+    force: yes
+
+- name: Add the self-signed certificate to your list of trusted certificates
+  shell: update-ca-trust
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply tasks/setup_registry.yaml]: Done

**Pull request merged**